geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/RedisCommandType.java

```diff
@@ -154,207 +165,274 @@
 
   /************* Cluster *****************/
-  CLUSTER(new ClusterExecutor(), SUPPORTED, new ClusterParameterRequirements()),
+  CLUSTER(new ClusterExecutor(), SUPPORTED, new Parameter().min(2)
+      .custom(ClusterParameterRequirements.checkParameters()).firstKey(0)
+      .flags(admin, random, stale)),
 
   /***************************************
    ********* Internal Commands ***********
    ***************************************/
   // do not call these directly, only to be used in other commands
-  INTERNALPTTL(new UnknownExecutor(), INTERNAL, new ExactParameterRequirements(2)),
-  INTERNALTYPE(new UnknownExecutor(), INTERNAL, new ExactParameterRequirements(2)),
-  INTERNALSMEMBERS(new UnknownExecutor(), INTERNAL, new ExactParameterRequirements(3)),
+  INTERNALPTTL(new UnknownExecutor(), INTERNAL, new Parameter().exact(2)),
+  INTERNALTYPE(new UnknownExecutor(), INTERNAL, new Parameter().exact(2)),
+  INTERNALSMEMBERS(new UnknownExecutor(), INTERNAL, new Parameter().exact(3)),
 
   /***************************************
    ******** Unsupported Commands *********
    ***************************************/
 
   /*************** Connection *************/
 
-  SELECT(new SelectExecutor(), UNSUPPORTED, new ExactParameterRequirements(2)),
+  SELECT(new SelectExecutor(), UNSUPPORTED, new Parameter().exact(2).firstKey(0).flags(loading, stale, fast)),
 
   /*************** Keys ******************/
 
-  SCAN(new ScanExecutor(), UNSUPPORTED,
-      new EvenParameterRequirements(ERROR_SYNTAX).and(new MinimumParameterRequirements(2))),
-  UNLINK(new DelExecutor(), UNSUPPORTED, new MinimumParameterRequirements(2)),
+  SCAN(new ScanExecutor(), UNSUPPORTED, new Parameter().min(2).even(ERROR_SYNTAX).firstKey(0).flags(readonly, random)),
+  UNLINK(new DelExecutor(), UNSUPPORTED, new Parameter().min(2).lastKey(-1).flags(write, fast)),
 
   /************** Strings ****************/
 
-  BITCOUNT(new BitCountExecutor(), UNSUPPORTED, new MinimumParameterRequirements(2)),
-  BITOP(new BitOpExecutor(), UNSUPPORTED, new MinimumParameterRequirements(4)),
-  BITPOS(new BitPosExecutor(), UNSUPPORTED, new MinimumParameterRequirements(3)),
-  GETBIT(new GetBitExecutor(), UNSUPPORTED, new ExactParameterRequirements(3)),
+  BITCOUNT(new BitCountExecutor(), UNSUPPORTED, new Parameter().min(2).flags(readonly)),
+  BITOP(new BitOpExecutor(), UNSUPPORTED,
+      new Parameter().min(4).firstKey(2).lastKey(-1).flags(write, denyoom)),
+  BITPOS(new BitPosExecutor(), UNSUPPORTED, new Parameter().min(3).flags(readonly)),
+  GETBIT(new GetBitExecutor(), UNSUPPORTED, new Parameter().exact(3).flags(readonly, fast)),
   MSETNX(new MSetNXExecutor(), UNSUPPORTED,
-      new MinimumParameterRequirements(3).and(new OddParameterRequirements())),
-  SETBIT(new SetBitExecutor(), UNSUPPORTED, new ExactParameterRequirements(4)),
+      new Parameter().min(3).odd().lastKey(-1).step(2).flags(write, denyoom)),
```

Comment:
What do you think of the idea of getting rid of "new Parameter()." by adding static factory methods to Parameter. You could name one of them min(int) and the other exact(int). It seems like you always start with min or exact. This is not a big deal but might make this class a bit more readable.

Reply:
That's a good idea, but it also seems that some of the deferred params use `odd()` and `custom()` which complicate things, so I'd prefer to leave it as it is unless you have some other suggestions.

Reply:
sounds good